src/serializer/ResidualHeapSerializer.js

```diff
@@ -1229,6 +1229,24 @@ export class ResidualHeapSerializer {
     return serializedValue;
   }
 
+  _serializeAbstractValue(val: AbstractValue): void | BabelNodeExpression {
+    invariant(val.kind !== "sentinel member expression", "invariant established by visitor");
+    if (val.hasIdentifier()) {
+      return this._serializeAbstractValueHelper(val);
+    } else {
+      // This abstract value's dependencies may have been declared
+      // but still need to check it again in case their serialized bodies are in different generator scope.
+      this.emitter.emitNowOrAfterWaitingForDependencies(val.args, () => {
+        const serializedValue = this._serializeAbstractValueHelper(val);
+        // Fetch the previous assigned identifier by residualHeapValueIdentifiers.setIdentifier()
+        // before _serializeAbstractValue() is called.
+        let uid = this.residualHeapValueIdentifiers.getIdentifierAndIncrementReferenceCount(val);
+        let declar = t.variableDeclaration("var", [t.variableDeclarator(uid, serializedValue)]);
+        this.emitter.emit(declar);
+      });
+    }
```

Comment:
Via a boolean flag, could we get an invariant that by the time `_serializeAbstractValue` returns, the function passed to `emitNowOrAfterWaitingForDependencies` must have been executed?

Reply:
Hmm, this invariant may not be true -- when this abstract value needs to wait for its generator body to be available, the function passed to emitNowOrAfterWaitingForDependencies() will be queued and executed later. 
This happens for the testcase together with this PR.

Reply:
I will land this for now and chat you offline tomorrow so that I can start send out other dependent PRs.